packages/pigeon/lib/objc_generator.dart

```diff
@@ -953,47 +945,55 @@ static id GetNullableObjectAtIndex(NSArray* array, NSInteger key) {
       });
     }
 
-    void writeFromMap() {
-      indent.write('+ ($className *)fromMap:(NSDictionary *)dict ');
+    void writeFromList() {
+      indent.write('+ ($className *)fromList:(NSMutableArray *)list ');
       indent.scoped('{', '}', () {
         const String resultName = 'pigeonResult';
         indent.writeln('$className *$resultName = [[$className alloc] init];');
-        for (final NamedType field in klass.fields) {
+        klass.fields
+            .toList()
+            .asMap()
+            .forEach((int index, final NamedType field) {
           if (enumNames.contains(field.type.baseName)) {
             indent.writeln(
-                '$resultName.${field.name} = [${_dictGetter(classNames, 'dict', field, options.prefix)} integerValue];');
+                '$resultName.${field.name} = [${_listGetter(classNames, 'list', field, index, options.prefix)} integerValue];');
           } else {
             indent.writeln(
-                '$resultName.${field.name} = ${_dictGetter(classNames, 'dict', field, options.prefix)};');
+                '$resultName.${field.name} = ${_listGetter(classNames, 'list', field, index, options.prefix)};');
             if (!field.type.isNullable) {
               indent
                   .writeln('NSAssert($resultName.${field.name} != nil, @"");');
             }
           }
-        }
+        });
         indent.writeln('return $resultName;');
       });
+
       indent.writeln(
-          '+ (nullable $className *)nullableFromMap:(NSDictionary *)dict { return (dict) ? [$className fromMap:dict] : nil; }');
+          '+ (nullable $className *)nullableFromList:(NSMutableArray *)list { return (list) ? [$className fromList:list] : nil; }');
     }
 
-    void writeToMap() {
-      indent.write('- (NSDictionary *)toMap ');
+    void writeToList() {
+      indent.write('- (NSMutableArray *)toList ');
       indent.scoped('{', '}', () {
-        indent.write('return');
```

Comment:
Since you'll be returning an `NSArray *`, you can use an array literal instead of populating dynamically, keeping the same structure as the dictionary code.